releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+
+Some companies may want to upgrade apps faster and more often than Kubernetes and operating system, because the former is less impactful to their workload than the latter and they do not require the latest Kubernetes and OS versions, but are satisfied with any supported and stable version. Others may not care about it and they would like to get newer versions of everything equally fast, or they want the latest Kubernetes releases to be deployed as soon as possible.
+
+To accommodate the needs of different types of businesses, we may need not one, but at least few different **release models**, so that customers can choose which one is working the best for their use case.
+
+In addition to all the above, even for a single release model, it may not be a good idea to make new releases available to all clusters at the same time, so it can be useful to have different **release channels**, where some release channels are getting new releases faster than the others. E.g. see [GKE release channels](https://cloud.google.com/kubernetes-engine/docs/concepts/release-channels). A development cluster may use a faster release channel, while a production cluster may use a slower one. Some companies may use faster release channel for all their clusters, while others may use a slower one similarly for all their workload.
+
+## 4. Proposal
+
+This section defines how releases are implemented, where they are, how they are developed, tested and delivered.
+
+Then it compares the proposed solution to the current one that we have for our new KaaS product that is based on Cluster API.
+
+### 4.1. Implementation of releases
+
+This document proposes to continue using `giantswarm/releases` repository for creating and delivering releases, albeit in a simplified way when compared to the vintage releases. Release resources are used in a minimal way, and cluster-$provider apps continue to be used for deploying workload clusters.
+
+Briefly put, cluster-$provider apps are still deployed in almost exactly same way, with the following few differences:
+- In cluster-$provider app manifest, instead of specifying `.spec.version`, we MUST specify `release.giantswarm.io/version` label.
```

Comment:
What's the benefit of using a label rather part of the spec? Labels can't have OpenAPI validation performed by the api-server.
Using a label for this seems like a hack and would require changes to our app platform to allow not specifying a version but only for cluster apps.

Reply:
+1, however, from batman times I remember that there's sometimes efficiency gains for label usage in terms of operator implementiations, but I'd say that should not result in omitting the spec field, but rather automatically adding/updating the label.

Reply:
Cluster apps are still just Apps (Giant Swarm app platform App resource) and we apply those to deploy all required resources for the workload cluster. That does not change here.

We will still have to apply real cluster-aws version. Release version is different than cluster-aws version. Also multiple releases can use the same cluster-aws version.

Today in App resources we don't have anything related to releases, so not sure we would put Release version in the App CR.

> Using a label for this seems like a hack and would require changes to our app platform to allow not specifying a version but only for cluster apps.

The workflow is the following and one change is needed in the app platform:
1. We deploy App CR with `release.giantswarm.io/version` label, e.g. "release.giantswarm.io/version: 25.0.0". It can be an annotation, or a new field in the spec, it doesn't matter, I was going for the simplest initial solution that works (it just cannot be `.spec.version` because cluster-aws version must be set there).
2. app-admission-controller mutates the App CR to set the correct `.spec.version`
  - it does this only for cluster apps (checking if an App is a cluster app is straightforward)
  - it reads release version from the label and it fetches the Release CR
  - it reads cluster-$provider app version from the Release CR
  - it sets App's `.spec.version`
3. cluster-$provider app got `.spec.version` set and it is deployed regularly like until now.

Reply:
Semi-related but it's just made me think - how would a change to the release version label trigger all the needed changes if the cluster app version remains the same? What would be responsible for updating the kubernetes version or flatcar version values for example?

Reply:
> Semi-related but it's just made me think - how would a change to the release version label trigger all the needed changes if the cluster app version remains the same?

Good question, gotta check how it behaves on the update (so far I successfully tested creation only).

Label update will definitely trigger App CR reconciliation, just not sure if that will re-render the templates (re-rendering of the templates would update e.g. OS version, k8s version, app versions).

In case that release label changes do not trigger re-rendering of the cluster-$provider app, an alternative solution would be to put the release version into cluster-$provider app's config (as changes in the config do trigger cluster-$provider app re-rendering). I thought of this also, but went with the simpler option with release label on the App CR (as having release label on the CRs is something that was common in vintage as well).

Reply:
I'm not 100% sure (Honeybadger should be able to confirm) but I'm pretty sure that just the label changing on the App wouldn't be enough to cause the Chart to re-render. It would need a change to the spec or the values for that to happen.

Reply:
cc @piontec or someone in Honeybadger should know

Reply:
I think that App CR change (even label) will get reconciled into Chart CR right away. Then not sure if that is enough for the chart-operator to re-render the templates immediately.

I can easily move release version to Helm values if needed here.

In any case - the RFC is not proposing the best solution (I have more than few improvements in mind), but the smallest incremental improvement that enables us to use Release CRs and releases repo. So:
- Main goal is to be able to use Release CR and read k8s/OS/app versions from there.
- It minimises impact to customers. Upgrade of existing clusters would be just two changes (slightly different compared to a regular version bump):
  - Remove `spec.version` in cluster-$provider app manifest.
  - Set release version.
- Where exactly release version is put is not crucial, at least not initially. We can also improve this later.
- Changes to some parts of the app platform (specifically to app-admission-controller) are needed here simply because we are using App platform to deploy cluster-$provider "apps".
  - quotes above (for "apps") because these are not apps in the sense of controllers/operators, but just Helm charts that deploy required CAPI/CAPx resources,
  - IMO we should not use App platform for this (but maybe just Flux HelmReleases), but that would be a bigger change now.

Reply:
After looking into this again I have moved release version from cluster-$provider App label to a Helm value `global.release.version`.

Thanks for the feedback here @AverageMarcus and @puja108!

I believe it is better like this, so now:
- Change in cluster chart is a bit simpler. OTOH change in mutating webhook has one more more step as it gets release version from the ConfigMap, but this is IMO fine.
- We could also do some basic release version validation with JSON schema.
- Cleaner Helm code to use the release version in the templates, e.g. to set it as a label.
- I have tested the cluster upgrade, it works as expected :)

Reply:
Will update the RFC.

Reply:
> I have tested the cluster upgrade, it works as expected :)

with the `release.version` as a helm value, I see a problem when the upgrade includes a change in version in `cluster-<provider>` app.

I believe that because the `release.version` is now in a configmap, the app-admission-controller won't update the `cluster-<provider>` app version until the actual app CR is modified.

Reply:
App CRs have ConfigMap version as the annotation. IIUC this gets updated when you change the app config (e.g. when you change the release version), so this App CR update should trigger the mutating webhook which then updates the App version.

```
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  annotations:
    app-operator.giantswarm.io/latest-configmap-version: "271755519"
```

Reply:
I missed that annotation. So all clear from my end. thanks for digging that up!